pkg/sampling/common.go

```diff
@@ -0,0 +1,127 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package sampling // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/sampling"
+
+import (
+	"errors"
+	"io"
+	"strings"
+
+	"go.uber.org/multierr"
+)
+
+// KV represents a key-value parsed from a section of the TraceState.
+type KV struct {
+	Key   string
+	Value string
+}
+
+var (
+	// ErrTraceStateSize is returned when a TraceState is over its
+	// size limit, as specified by W3C.
+	ErrTraceStateSize = errors.New("invalid tracestate size")
+)
+
+// keyValueScanner defines distinct scanner behaviors for lists of
+// key-values.
+type keyValueScanner struct {
+	// maxItems is 32 or -1
+	maxItems int
+	// trim is set if OWS (optional whitespace) should be removed
+	trim bool
+	// separator is , or ;
+	separator byte
+	// equality is = or :
+	equality byte
+}
+
+// commonTraceState is embedded in both W3C and OTel trace states.
+type commonTraceState struct {
+	kvs []KV
+}
+
+func (cts commonTraceState) HasExtraValues() bool {
+	return len(cts.kvs) != 0
+}
+
+func (cts commonTraceState) ExtraValues() []KV {
+	return cts.kvs
+}
+
+// trimOws removes optional whitespace on both ends of a string.
+func trimOws(input string) string {
+	// Hard-codes the value of owsCharset
+	for len(input) > 0 && (input[0] == ' ' || input[0] == '\t') {
+		input = input[1:]
+	}
+	for len(input) > 0 && (input[len(input)-1] == ' ' || input[len(input)-1] == '\t') {
+		input = input[:len(input)-1]
+	}
+	return input
+}
+
+// scanKeyValues is common code to scan either W3C or OTel tracestate
+// entries, as parameterized in the keyValueScanner struct.
+func (s keyValueScanner) scanKeyValues(input string, f func(key, value string) error) error {
+	var rval error
+	items := 0
+	for input != "" {
+		items++
+		if s.maxItems > 0 && items >= s.maxItems {
+			// W3C specifies max 32 entries, tested here
+			// instead of via the regexp.
+			return ErrTraceStateSize
+		}
+
+		sep := strings.IndexByte(input, s.separator)
+
+		var member string
+		if sep < 0 {
+			member = input
+			input = ""
+		} else {
+			member = input[:sep]
+			input = input[sep+1:]
+		}
+
+		if s.trim {
+			// Trim only required for W3C; OTel does not
+			// specify whitespace for its value encoding.
+			member = trimOws(member)
+		}
+
+		if member == "" {
+			// W3C allows empty list members.
+			continue
+		}
+
+		eq := strings.IndexByte(member, s.equality)
+		if eq < 0 {
+			// A regexp should have rejected this input.
```

Comment:
not sure i follow this comment here

Reply:
Changed. An early review by @kentquirk requested the use of regexps to validate syntax, which has been done. This logic re-parses the string instead of using the regexp match results, because it's simpler.

```
			// We expect to find the `s.equality`
			// character in this string because we have
			// already validated the whole input syntax
			// before calling this parser.  I.e., this can
			// never happen, and if it did, the result
			// would be to skip malformed entries.
```